src/EllCrv/MordellWeilRank.jl

```diff
@@ -0,0 +1,365 @@
+###############################################################################
+#
+#  Descent methods for rank computations and local solubility of quartics
+#
+#
+#  (C) 2022 Jeroen Hanselman
+#
+###############################################################################
+
+export  quartic_local_solubility, R_soluble, Qp_soluble, quartic_rational_point_search, rank_2_torsion
+
+###############################################################################
+#
+#  Quartic local solubility
+#
+###############################################################################
+
+#TODO: Extend this to hyperelliptic curves
+#TODO: Do this over number fields
+
+@doc Markdown.doc"""
+    quartic_local_solubility(a::fmpq, b::fmpq, c::fmpq, d::fmpq, e::fmpq) -> Bool
+
+Check if the quartic defined by ax^4+bx^3+cx^2+dx+e is soluble over the real field $\mathbb{R}$ and over all local fields $\mathbb{Q}_p$.
+"""
+function quartic_local_solubility(a, b, c, d, e)
+
+  if !(R_soluble(a,b,c,d,e))
+    return false
+  end
+
+  if !(Qp_soluble(a,b,c,d,e,2)) 
+    return false
+  end
+
+  R, x = PolynomialRing(QQ)
+  
+  delta = discriminant(a*x^4 + b*x^3 + c*x^2 +d*x + e)
+
+  fac = factor(numerator(delta))
+  p_list = [i[1] for i in fac]
+  p_list = sort(p_list)
+
+  for p in p_list
+    if (!Qp_soluble(a,b,c,d,e,p))
+      return false
+    end
+  end
+
+  return true
+end
+
+@doc Markdown.doc"""
+    R_soluble(a::fmpq, b::fmpq, c::fmpq, d::fmpq, e::fmpq) -> Bool
+
+Check if the quartic defined by $ax^4+bx^3+cx^2+dx+e$ has a solution over $\mathbb{R}$.
+"""
+function R_soluble(a, b, c, d, e)
+  if a>0
+    return true
+  end
+  
+  R,x = PolynomialRing(ZZ,"x")
+  
+  if (signature(a*x^4+b*x^3+c*x^2+d*x+e)[1]>0)
+    return true
+  else
+    return false
+  end
+end
+
+@doc Markdown.doc"""
+    Qp_soluble(a::fmpq, b::fmpq, c::fmpq, d::fmpq, e::fmpq,
+    p::fmpz) -> Bool
+
+Check if the quartic defined by $ax^4+bx^3+cx^2+dx+e$ has a solution over the local field $\mathbb{Q}_p$.
+"""
+function Qp_soluble(a, b, c, d, e, p)
+  R,x = PolynomialRing(QQ,"x")
+  if(Zp_soluble(a,b,c,d,e,0,p,0))
+    return true
+  end
+  
+  if(Zp_soluble(e,d,c,b,a,0,p,1))
+    return true
+  end
+  return false
+end
+
+
+function Zp_soluble(a,b,c,d,e,x_k,p,k)
+  if (p==2)
+    code = lemma7(a,b,c,d,e,x_k,k)
+  else
+    code = lemma6(a,b,c,d,e,x_k,p,k)
+  end
+
+  if code==1
+    return true
+  end
+  if code==-1
+    return false
+  end
+  
+  for t in (0:p-1)
+    if Zp_soluble(a,b,c,d,e,x_k+t*p^k,p,k+1)
+      return true
+    end
+  end
+
+  return false
+end
+
+#Z_p lifting for odd p
+
+function lemma6(a,b,c,d,e,x,p,n)
```

Comment:
This is inscrutable; at the very least add a comment indicating where this "Lemma 6" can be found?